src/coord/src/catalog/storage.rs

```diff
@@ -128,22 +148,80 @@ const MIGRATIONS: &[&str] = &[
     // Makes the information_schema schema literal so it can store functions.
     //
     // Introduced in v0.9.12.
-    "INSERT INTO schemas (database_id, name) VALUES
+    &"INSERT INTO schemas (database_id, name) VALUES
         (NULL, 'information_schema');",
     // Adds index to timestamp table to more efficiently compact timestamps.
     //
     // Introduced in v0.12.0.
-    "CREATE INDEX timestamps_sid_timestamp ON timestamps (sid, timestamp)",
+    &"CREATE INDEX timestamps_sid_timestamp ON timestamps (sid, timestamp)",
     // Adds table to track users' compute instances.
     //
     // Introduced in v0.22.0.
-    "CREATE TABLE compute_instances (
+    &"CREATE TABLE compute_instances (
         id   integer PRIMARY KEY,
         name text NOT NULL UNIQUE
     );
     INSERT INTO compute_instances VALUES (1, 'default');",
     // Introduced in v0.24.0.
-    "ALTER TABLE compute_instances ADD COLUMN config text",
+    &"ALTER TABLE compute_instances ADD COLUMN config text",
+    // Migrates timestamp bindings from the coordinator's catalog to STORAGE's internal state
+    // Introduced in v0.25.0.
+    &|data_dir_path: &Path, tx: &rusqlite::Transaction| {
+        let source_ids = tx
+            .prepare("SELECT DISTINCT sid FROM timestamps")?
+            .query_and_then([], |row| Ok(row.get::<_, SqlVal<GlobalId>>(0)?.0))?
+            .collect::<Result<Vec<_>, Error>>()?;
+
+        std::fs::create_dir_all(data_dir_path.join("controllers/storage/")).unwrap();
+        let stash = Stash::open(&data_dir_path.join("controllers/storage/timestamp_bindings"))
```

Comment:
I was thinking that we'd have one `Stash` per layer, and namespace collection IDs as necessary. So `controllers/storage` would have `timestamp-bindings-1`, `timestamp-bindings-2`, etc. That leaves open the possibility of adding transactions to `Stash`, and then allowing each layer to transactionally update any of its collections. (Adding cross-collection transactions seems straightforward; cross-`Stash` transactions are far fetched though.)